Create base collection layout and assets

ehelms opened the pull request:
Borrowed and minor adaptations from https://github.com/theforeman/foreman-ansible-modules to match as close as possible to test/release/management of our collections.

mdellweg commented on .github/ISSUE_TEMPLATE/bug_report.md:
````diff
@@ -0,0 +1,49 @@
+
+```
+
+##### APYPIE VERSION
````
If i understand correctly, this should be the version of FAM here.

mdellweg commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,99 @@
+      matrix:
+        python: ["2.7", "3.5", "3.6", "3.7"]
+        ansible:
+          - stable-2.7
```
This is going to be a collection to begin with, right?
Then you should not test ansible 2.7
However 2.10 would be interesting. ;)

mdellweg commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,99 @@
+        run: make SANITY_OPTS="--local" PYTHON_VERSION=${{ matrix.python }} sanity dist-test
+        if: matrix.ansible == 'devel'
+
+  checkmode:
```
Not sure this is a thing if we don't have modules here.

ehelms replied:
I think we will in the end, so I debated whether to leave it for when that time comes or just be ready for it.

mdellweg commented on Makefile:
```diff
@@ -0,0 +1,117 @@
+
+FORCE:
+
+.PHONY: help dist lint sanity test test-crud test-check-mode test-other setup test-setup FORCE
```
This reminds me that i wanted to look into reusable makefile includes.

mdellweg commented on docs/developing.md:
```diff
@@ -0,0 +1,101 @@
+# How to write modules
```
This whole file seems very specific to FAM.

mdellweg commented on docs/index.rst:
```diff
@@ -0,0 +1,28 @@
+   You can adapt this file completely to your liking, but it should at least
+   contain the root `toctree` directive.
+
+Welcome to Foreman Ansible Modules's documentation!
```
... Operations ...

mdellweg commented on docs/releasing.md:
```diff
@@ -0,0 +1,14 @@
+* After merging, tag the merge commit with `v<version number>` (i.e. v0.1.1).
+  Please use signed tags.
+
+Consider filing a packaging PR on [foreman-packaging](https://github.com/theforeman/foreman-packaging).
```
:D This file seems generic enough.

mdellweg commented on docs/testing.md:
```diff
@@ -0,0 +1,103 @@
+# Testing Foreman Ansible Modules
```
Again, (in parts) very FAM specific

mdellweg reviewed (commented):
This is by far not a complete review. i.e. the whole docs processing is a black box to me.
However, while all looks quite familiar, can it be, that you copied the files from a rather old instantiation of FAM?

ehelms commented:
@mdellweg bah you are right, I was on the deprecated master branch when I copied (@evgeni can we delete the master branch from FAM to avoid confusion?) and I will update this based on latest develop

evgeni commented:
@ehelms I did long ago: https://github.com/theforeman/foreman-ansible-modules/branches (and closed some PRs doing so **cough**)

ehelms commented:
@evgeni wth, figured my script would have errored trying to fetch it, this is why you shouldn't let me play with Ansible code repos :)

evgeni commented:
Well, if your local checkout still had a `master`, it might have taken that? Or was it a fresh clone?

ehelms commented:
Latest update, I have pushed empty github workflows to ensure what gets merged here works in some fashion from the outset. I have updated the Makefile to account for this and adjusted the test templates as well.

mdellweg commented on .github/ISSUE_TEMPLATE/bug_report.md:
````diff
@@ -0,0 +1,51 @@
+
+```
+
+##### COLLECTION VERSION
````
:+1:

mdellweg commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,126 @@
+          - "3.7"
+          - "3.8"
+        ansible:
+          - stable-2.8
```
Are you sure, you want to start a new collection with ansible 2.8 in mind?

ehelms replied:
Dropped 2.8. Starting clean.

mdellweg commented on Dockerfile:
```diff
@@ -0,0 +1,8 @@
+
+RUN git clone https://github.com/theforeman/foreman-operations-collection ~/.ansible/plugins/modules/
+
+RUN cd ~/.ansible/plugins/modules && pip install -r requirements-dev.txt
```
Are we using this container anywhere?

mdellweg commented on README.md:
```diff
@@ -1,2 +1,50 @@
+
+### Supported Python Versions
+
+Starting with Ansible 2.7, Ansible only supports Python 2.7 and 3.5 (and higher). These are also the only Python versions we develop and test the modules against.
```
2.7 (Ansible) is definitely not supported anymore...

ehelms replied:
Agreed, I do not think this is intended to say that. I think is this just giving history that in Ansible 2.7 they started only supporting Python 2.7 and 3.5 (or higher). That is, putting the blame on Ansible. I can just reword this :)

mdellweg commented on README.md:
```diff
@@ -1,2 +1,50 @@
+
+### Supported Ansible Versions
+
+The supported Ansible versions are aligned with currently maintained Ansible versions. You can find the list of maintained Ansible versions [here](https://docs.ansible.com/ansible/latest/reference_appendices/release_and_maintenance.html#release-status). Older Ansible versions (Ansible >= 2.3) should work but are not tested.
```
2.3 ???
This is either >= 2.8 or >=2.9, because collections.

mdellweg commented on README.md:
```diff
@@ -1,2 +1,50 @@
+
+## Installation
+
+There are currently two ways to use the modules in your setup: install from Ansible Galaxy or via RPM.
```
I can imagine RPM is planned...

ehelms replied:
Correct.

ehelms replied:
But I can rip all of that out and until I get that far if y'all prefer.

mdellweg commented on README.md:
```diff
@@ -1,2 +1,50 @@
+
+After the installation, the assets are available as `theforeman.operations.<name>`. Please see the [Using Ansible collections documentation](https://docs.ansible.com/ansible/devel/user_guide/collections_using.html) for further details.
+
+### Installation via RPM
```
In this section at least the numbers are wrong.

mdellweg reviewed (commented):
Found some more inconsistencies...

mdellweg commented:
If not supporting 2.8, you can rip out all the mazer hacks.

ehelms commented:
> If not supporting 2.8, you can rip out all the mazer hacks.

Great point, deleted them.

mdellweg commented on requirements-common.txt:
```diff
@@ -0,0 +1,2 @@
+ansible>=2.3
```
```suggestion
ansible>=2.9
```

evgeni replied:
`requirements-common` doesn't exist in FAM anymore, and we don't define a dependecy on ansible besides in the docs.

mdellweg reviewed (approved):
I cannot speak so much about the doc generation process. @evgeni ?

But in general, i think this is something we can have a good start from.

evgeni commented on docs/_gh_include/header.inc:
```diff
@@ -0,0 +1,17 @@
+  </head>
+
+  <body style="margin:0px 15px">
+    <h1>Foreman Ansible Modules</h1>
```
wrong name in that file ;)

evgeni commented on docs/_templates/plugin.rst.j2:
```diff
@@ -0,0 +1,469 @@
+:source: @{ source }@
```
that file should not be there anymore

evgeni commented on docs/_themes/ansible_rtd_theme/static/ansible_rtd.css:
```diff
@@ -0,0 +1,515 @@
+@import url('css/theme.css');
```
that folder (`_themes`) should not be there anymore

evgeni commented on docs/index.rst:
```diff
@@ -0,0 +1,32 @@
+   You can adapt this file completely to your liking, but it should at least
+   contain the root `toctree` directive.
+
+Welcome to Foreman Operations Collection' documentation!
```
```suggestion
Welcome to Foreman Operations Collection documentation!
```

evgeni commented on docs/index.rst:
```diff
@@ -0,0 +1,32 @@
+   :caption: Developer documentation
+
+   developing
+   api
```
I don't think we have an `api` file yet.

evgeni commented on galaxy.yml:
```diff
@@ -0,0 +1,17 @@
+namespace: "theforeman"
+name: "operations"
+description: Ansible assets to manage Foreman operations such as install, upgrade and maintenance tasks
```
starting a description of an ansible collection with "ansible assets" sounds kinda repetitive?

```suggestion
description: Foreman operations tasks such as install, upgrade and maintenance
```

ehelms commented:
Comments addressed, I appreciate the patience reviewing a big PR to get things started.

evgeni commented:
@ehelms i meant the full docs/_themes and docs/_templates folders can go, they aren't present in the current FAM code.

ehelms commented:
> @ehelms i meant the full docs/_themes and docs/_templates folders can go, they aren't present in the current FAM code.

You should have just said that :) THEY GONE

evgeni commented:
I thought I did!

evgeni commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,114 @@
+        run: pip install --upgrade git+https://github.com/ansible/ansible.git@${{ matrix.ansible }}
+      - name: Install dependencies
+        run: make test-setup
+      - name: fake redhat-uep.pem for redhat_manifest module
```
This step won't be necessary here

evgeni commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,114 @@
+        uses: actions/upload-artifact@v2
+        with:
+          name: collection
+          path: theforeman-foreman-*.tar.gz
```
This path needs updating for the new name

evgeni commented on README.md:
```diff
@@ -1,2 +1,50 @@
+
+The supported Ansible versions are aligned with currently maintained Ansible versions. You can find the list of maintained Ansible versions [here](https://docs.ansible.com/ansible/latest/reference_appendices/release_and_maintenance.html#release-status).
+
+As we're using Ansible's [documentation fragment](https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html#documentation-fragments) feature, that was introduced in Ansible 2.8, `ansible-doc` prior to 2.8 won't be able to display the module documentation, but the modules will still run fine with `ansible` and `ansible-playbook`.
```
This section is irrelevant if we only support 2.9+

evgeni commented on README.md:
```diff
@@ -1,2 +1,50 @@
+
+### Installation from Ansible Galaxy
+
+You can install the collection from [Ansible Galaxy](https://galaxy.ansible.com/theforeman/foreman) by running `ansible-galaxy collection install theforeman.operations`.
```
Wrong galaxy link ;)

evgeni commented:
Thanks @ehelms!

I think the only two remaining points are:
* what is this `Dockerfile` used for? We don't have one in FAM.
* `requirements-common.txt` should go away as in FAM

ehelms commented:
> I think the only two remaining points are:
> 
> * what is this `Dockerfile` used for? We don't have one in FAM.
> * `requirements-common.txt` should go away as in FAM

Removed
